src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -48,60 +49,99 @@ internal static SafeChannelBindingHandle QueryChannelBinding(SafeSslHandle conte
 
         internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX509Handle certHandle, SafeEvpPKeyHandle certKeyHandle, EncryptionPolicy policy, bool isServer, bool remoteCertRequired)
         {
-            SafeSslHandle context = null;
+            bool hasCertReference = false, 
+                hasCertKeyReference = false;
 
-            IntPtr method = GetSslMethod(protocols);
+            bool hasCertificateAndKey = 
+                certHandle != null && !certHandle.IsInvalid
+                && certKeyHandle != null && !certKeyHandle.IsInvalid;
 
-            using (SafeSslContextHandle innerContext = Ssl.SslCtxCreate(method))
+            try
             {
-                if (innerContext.IsInvalid)
+                if (hasCertificateAndKey)
                 {
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    certHandle.DangerousAddRef(ref hasCertReference);
+                    certKeyHandle.DangerousAddRef(ref hasCertKeyReference);
```

Comment:
Why are these add-refs being done way up here?  The try/finally can be narrowed to just the area that's using DangerousGetHandle.  Our goal here isn't to throw a particular exception.  It's to ensure that if the handles are incorrectly disposed of while they're in use, that we don't corrupt state, free-while-in-use, access freed state, etc.